pywbem_mock/_inmemoryrepository.py

```diff
@@ -0,0 +1,319 @@
+    def get(self, name, copy=True):
+        """
+        """
+        #print("GET %s" % name)
+        if name in self._data:
+            if copy:
+                return self._data[name].copy()
+            else:
+                return self._data[name]
+        else:
+            raise KeyError('{} not in datastore'.format(name))
+
+    def create(self, name, cim_object):
+        """
+        """
+        #print("CREATE %s %s " % (name, self._cim_object_type))
+        assert isinstance(cim_object, self._cim_object_type)
+        if name in self._data:
+            #print('CREATE already in datastoe')
+            raise ValueError('{} already in datastore'.format(name))
+        else:
+            self._data[name] = cim_object
+
+    def update(self, name, cim_object):
+        """
+        """
+        assert isinstance(cim_object, self._cim_object_type)
+
+        if name not in self._data:
+            raise KeyError('{} not in datastore'.format(name))
+
+    def delete(self, name):
+        """
+        """
+        if name in self._data:
+            del self._data[name]
+        else:
+            raise KeyError('{} not in datastore'.format(name))
+
+    def names(self):
+        """
+        Get the names for all of the objects in data store. In the case
+        of case insensitive store, returns the real names
+
+        Returns:
+            list of the names of the objects in the data stoe
+
+        Raises: TODO
+        """
+        return list(six.iterkeys(self._data))
+
+    def iter(self):
+        """
+        Returns iterator for the data defined in the data store
+
+        """
+        return six.itervalues(self._data)
+
+    def len(self):
+        """
+        Returns count of objects in the object store
+
+        """
+        return len(self._data)
+
+
+class InMemoryRepository(BaseRepository):
+    """
+    Defines the data store and access methods for a simple CIMClass,
+    CIMInstance, and CIMQualifierDeclaration repository that maintains the date
+    in memory.
+
+    The API for this data store is defined in the BaseRepository class and
+    the BaseObjStore class.
+    """
+    def __init__(self, initial_namespace):
+        """
+        Initialize the InMemoryRepository by creatiing the initial
+        namespace definition and the data stores for CIM classes, CIM instances,
+        and CIM qualifier declarations.
+
+          Parameters:
+             Initial namespace defined for this repository
+        """
+        self._repository = NocaseDict()
+
+        namespace = initial_namespace.strip('/')
+        self._repository[namespace] = {}
+        self._repository[namespace]['classes'] = InMemoryObjStore(
+            True, CIMClass, self._repository)
+
+        self._repository[namespace]['instances'] = InMemoryObjStore(
+            False, CIMInstance, self._repository)
+        self._repository[namespace]['qualifiers'] = InMemoryObjStore(
+            True, CIMQualifierDeclaration, self._repository)
```

Comment:
* Why an initial namespace? From a functional perspective, the repository can live well in the state where it has zero namespaces.
* If we really want to have an initial namespace in the init method, it should initialize the repo to empty namespaces and use the add_namespace method to add it, to avoid duplicating the code.

Reply:
1. The  initial namespace corresponds to the concept that WBEMConnection has an initial namespace.  I could not figure out any condition where we would create initialize a repository without an initial namespace

2. I agree it should go through add_namespace. 

DONE for next commit